packages/muelu/src/Smoothers/MueLu_Ifpack2Smoother_def.hpp

```diff
@@ -447,6 +466,58 @@ namespace MueLu {
     prec_->compute();
   }
 
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  void Ifpack2Smoother<Scalar, LocalOrdinal, GlobalOrdinal, Node>::SetupBlockRelaxation(Level& currentLevel) {
+    typedef Tpetra::RowMatrix<SC,LO,GO,NO> tRowMatrix;
+
+    RCP<BlockedCrsMatrix> bA = rcp_dynamic_cast<BlockedCrsMatrix>(A_);
+    if (!bA.is_null())
+      A_ = bA->Merge();
+
+    RCP<const tRowMatrix> tA = Utilities::Op2NonConstTpetraRow(A_);
+
+    bool reusePreconditioner = false;
+    if (this->IsSetup() == true) {
+      // Reuse the constructed preconditioner
+      this->GetOStream(Runtime1) << "MueLu::Ifpack2Smoother::SetupGeneric(): Setup() has already been called, assuming reuse" << std::endl;
+
+      RCP<Ifpack2::Details::CanChangeMatrix<tRowMatrix> > prec = rcp_dynamic_cast<Ifpack2::Details::CanChangeMatrix<tRowMatrix> >(prec_);
+      if (!prec.is_null()) {
+#ifdef IFPACK2_HAS_PROPER_REUSE
+        prec->resetMatrix(tA);
+        reusePreconditioner = true;
+#else
+        this->GetOStream(Errors) << "Ifpack2 does not have proper reuse yet." << std::endl;
+#endif
+
+      } else {
+        this->GetOStream(Warnings0) << "MueLu::Ifpack2Smoother::SetupSchwarz(): reuse of this type is not available (failed cast to CanChangeMatrix), "
+            "reverting to full construction" << std::endl;
+      }
+    }
+
+    if (!reusePreconditioner) {
+      ParameterList& myparamList = const_cast<ParameterList&>(this->GetParameterList());
+      std::cout << "*** print preconditioner parameter list from setup ***" << std::endl;
+      myparamList.print();
+      if(myparamList.isParameter("partitioner: type") &&
+         myparamList.get<std::string>("partitioner: type") == "line") {
+        Teuchos::RCP<Xpetra::MultiVector<double,LO,GO,NO> > xCoordinates =
```

Comment:
The code would be easier to read if you start using `auto`.

Reply:
@aprokop this is a question of style, I do not think that the use of auto is making the code clearer.
I have been debugging code that uses auto a lot and I find that I am spending more time now trying to understand what variable types are than I use to. Plus auto with template is usually a recipe for disaster especially when you have a special instantiation on `double` instead of the usual `Scalar`.
Again I think this is style...